### DISCUSSION TITLE
Extending tsconfig options

### DISCUSSION THREAD

**Discussion by janeklb** [projen/projen#2982] [Q&A] [answered]
I need to add some properties to my `tsconfig.json` file that are not available in `TypescriptConfigOptions`.

I can achieve the desired end-result by manually creating a ts config "extension" file (eg.  `tsconfig.ext.json`), including all of the additional/required config properties, and referencing it via the projen-supported `extends` option. 
```ts
tsconfig: {
  extends: javascript.TypescriptConfigExtends.fromPaths(["./tsconfig.ext.json"]),
}
```

While this works (though it feels like a _slight_ misuse of the `extends` tsconfig setting), I'm interested to get a maintainer's take on whether it makes sense to update `projen`'s `TypescriptConfigOptions` to support more of [tsconfig](https://www.typescriptlang.org/tsconfig)? This is a can of worms that I haven't seen opened anywhere else,.. but better sooner than later, i guess! 

Related: #2983

**Comment by mrgrain** [comment, marked as answer]
Hi, thanks for you excellent question! I agree using `extends` for this feels like a misuse of the feature. 

Thankfully projen has first class support for low-level operations on arbitrary object files.  See [documentation for details](https://projen.io/escape-hatches.html). To change your TypeScript config, you can do the following to set any value:

```ts
project.tsconfig?.file.addOverride()
project.tsconfig?.file.patch()
```

> I'm interested to get a maintainer's take on whether it makes sense to update projen's TypescriptConfigOptions to support more of [tsconfig](https://www.typescriptlang.org/tsconfig)? This is a can of worms that I haven't seen opened anywhere else,.. but better sooner than later, i guess!

Yes! Would love to receive a PR for missing options. It's a bit of a can of worms, but we have absolutely done this in the past.

**Reply by janeklb** [reply]
Thank you for a link to the overrides / escape hatches -- i had only glossed over this while reading docs and forgot to check it out again.

> Yes! Would love to receive a PR for missing options. It's a bit of a can of worms, but we have absolutely done this in the past.

So.. i spoke too soon wrt. updating projen's `tsconfig` support. Turns out the stuff I needed is actually not part of TypeScript, but actually `ts-node` (see their [docs](https://www.npmjs.com/package/ts-node#via-tsconfigjson-recommended)). Would be happy to contribute support for that if you can suggest a way forward?

I guess it would probably not be an addition to `tsconfig` / `TypescriptConfigOptions`, but rather a new field? eg. `tsconfigTsNode?: TsNodeTypescriptConfigOptions`

**Reply by mrgrain** [reply]
Ah interesting. This might be a separate `tsNodeConfig` option and a `TsNode` component that will add itself to the `TypescriptConfig` maybe? 🤔

**Reply by janeklb** [reply]
Thanks @mrgrain. I'd be happy to try and contribute if you could provide a few more pointers.
